kafka/parsing.go

```diff
@@ -0,0 +1,183 @@
+package kafka
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	exporter "github.com/kawamuray/prometheus-kafka-consumer-group-exporter"
+	"github.com/prometheus/common/log"
+)
+
+func parseGroups(output CommandOutput) ([]string, error) {
+	if strings.Contains(output.Stderr, "java.lang.RuntimeException") {
+		return nil, fmt.Errorf("Got runtime error when executing script. Output: %s", output)
+	}
+
+	lines := strings.Split(output.Stdout, "\n")
+	groups := make([]string, 0, len(lines))
+	for _, line := range lines {
+		if line != "" {
+			groups = append(groups, line)
+		}
+	}
+
+	return groups, nil
+}
+
+type regexpParser struct {
+	// header is the expected format of the first line. Used as fingerprint to distinguish if this parser is the right one.
+	header *regexp.Regexp
+	// line is the regexp used for the remaining lines of the output (as long as Header matched correctly).
+	line *regexp.Regexp
+	//
+	indexByName map[string]int
+}
+
+func newRegexpParser(header, line *regexp.Regexp) (*regexpParser, error) {
+	indexByName := make(map[string]int)
+	for i, name := range line.SubexpNames() {
+		indexByName[name] = i
+	}
+
+	for _, field := range []string{"topic", "partitionId", "currentOffset", "lag", "clientId", "consumerAddress"} {
+		if _, exists := indexByName[field]; !exists {
+			return nil, fmt.Errorf("line regexp missing '%s' capturing group", field)
+		}
+	}
+
+	return &regexpParser{header.Copy(), line.Copy(), indexByName}, nil
+}
+
+func (p *regexpParser) Parse(output CommandOutput) ([]exporter.PartitionInfo, error) {
+	lines := strings.Split(output.Stdout, "\n")
+	if len(lines) == 0 {
+		return nil, errors.New("empty output. stderr: " + output.Stderr)
+	}
+	headerLine := lines[0]
+	dataLines := lines[1:]
+
+	if !p.header.MatchString(headerLine) {
+		return nil, errors.New("incorrect header")
+	}
+
+	partitions := make([]exporter.PartitionInfo, 0, len(dataLines))
+	var err error
+	var partition *exporter.PartitionInfo
+	for _, line := range dataLines {
+		partition, err = p.parseLine(line)
+		if err != nil {
+			break
+		}
+		partitions = append(partitions, *partition)
+	}
+	return partitions, err
+}
+
+func (p *regexpParser) String() string {
+	return fmt.Sprintf("regexpParser{header: `%s`, line: `%s`}", p.header, p.line)
+}
+
+func (p *regexpParser) parseLine(line string) (*exporter.PartitionInfo, error) {
+	matches := p.line.FindStringSubmatch(line)
+
+	var err error
+
+	var currentOffset int64
+	currentOffset, err = parseLong(matches[p.indexByName["currentOffset"]])
```

Comment:
This will crash with the 0.10.1 format:

```
panic: runtime error: index out of range

goroutine 40 [running]:
github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka.(*regexpParser).parseLine(0xc4200b0f20, 0xc420186000, 0x0, 0xc4201b47d0, 0x0, 0x0)
	/src/github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka/parsing.go:89 +0x654
github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka.(*regexpParser).Parse(0xc4200b0f20, 0xc420186000, 0x3177, 0x0, 0x0, 0x0, 0x0, 0x0, 0xad2180, 0xc42012a860)
	/src/github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka/parsing.go:70 +0x271
github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka.(*DelegatingParser).Parse(0xc420112ac0, 0xc420186000, 0x3177, 0x0, 0x0, 0x3, 0xc420186000, 0x3177, 0x0, 0x0)
	/src/github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka/parsing.go:172 +0x8f
github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka.(*ConsumerGroupsCommandClient).DescribeGroup(0xc420100960, 0xad8d80, 0xc42013ccc0, 0xc42012c8a0, 0x12, 0x0, 0x0, 0x0, 0x0, 0x0)
	/src/github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka/collector.go:79 +0x149
github.com/kawamuray/prometheus-kafka-consumer-group-exporter/sync.(*FanInConsumerGroupInfoClient).describeLoop.func1(0xc4200116c0, 0xc42012c8a0, 0x12, 0xc42013cd80, 0xad8d80, 0xc42013ccc0, 0xc42013ce40)
	/src/github.com/kawamuray/prometheus-kafka-consumer-group-exporter/sync/metrics.go:159 +0x71
created by github.com/kawamuray/prometheus-kafka-consumer-group-exporter/sync.(*FanInConsumerGroupInfoClient).describeLoop
	/src/github.com/kawamuray/prometheus-kafka-consumer-group-exporter/sync/metrics.go:164 +0x178
```

Reply:
Thanks! Would it be possible for you to send me a sample of the output from the `kafka-consumer-groups.sh` script of `0.10.1.X` release? If you could separate `stderr` from `stdout` that would be nice.

Reply:
Sure:

```sh
# ./kafka-consumer-groups.sh --describe --group group1 --bootstrap-server bootstrap:9092
GROUP                          TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             OWNER
group1             requests                       11         1176417636761   1176419294709   1657948         consumer1_/1.1.1.5
group1             requests                       25         1144248140115   1144249685348   1545233         consumer1_/1.1.1.18
group1             requests                       24         1101749049926   1101751118906   2068980         consumer1_/1.1.1.23
group1             requests                       16         1065240411131   1065242284398   1873267         consumer1_/1.1.1.6
group1             requests                       81         76858987875     76860609239     1621364         consumer1_/1.1.1.18
group1             requests                       68         1182773827958   1182897904075   124076117       consumer1_/1.1.1.19
group1             requests                       66         1058973733081   1058974821719   1088638         consumer1_/1.1.1.12
group1             requests                       30         1156698657130   1156700015952   1358822         consumer1_/1.1.1.10
...
```

Reply:
@vavrusa Thanks a lot for that test case. I believe we now also have support for 0.10.1.X 🎉